macOS: hickory parse failure on link-local IPv6 nameserver causes opaque DNS timeout with no fallback (cross-ref #1700, #1699, hickory-dns#3667)

> **Disclaimer**: this report was drafted by an automated AI agent
> (@claude / Claude Code) while diagnosing a downstream user report
> with the end user in the loop. The diagnosis, MRE, trace excerpts,
> and system facts below were collected from a real reproduction on
> the end user's machine. The end user has reviewed this report
> before submission.

**Describe the bug**

On macOS, when `/etc/resolv.conf` / SCDynamicStore contains any IPv6
link-local nameserver with a zone identifier (e.g.
`fe80::1%en1`) — which is the default macOS behaviour any time the
upstream router advertises IPv6 DNS via Router Advertisement RDNSS —
`cargo binstall` fails every request with an opaque DNS timeout, even
though the user's system DNS works perfectly for every other tool
(`dig`, `curl`, Safari, `cargo` itself, etc.).

The root cause is upstream in `hickory-resolver` 0.26.1's macOS
system-config loader — reported as
[hickory-dns/hickory-dns#3667](https://github.com/hickory-dns/hickory-dns/issues/3667).
A single unparseable nameserver line causes `read_system_conf()` to
discard the entire configuration. cargo-binstall then falls back to
hard-coded DoH3 resolvers (Quad9 / Cloudflare / Google), and on the
reproducing machine those fail with `invalid peer certificate:
UnknownIssuer` (a separate problem in the same vicinity as
[hickory-dns#2066](https://github.com/hickory-dns/hickory-dns/issues/2066)),
so the user just sees:

```
ERROR Fatal error:
  × For crate just-lsp: could not GET https://index.crates.io/config.json:
  │ error sending request for url (https://index.crates.io/config.json)
  ...
  ├─▶ client error (Connect)
  ├─▶ dns error
  ╰─▶ request timed out
```

Filing here because — although the upstream hickory parse bug is the
root cause — cargo-binstall has several existing tracking issues
covering the same failure pattern (a single hickory misbehaviour
takes the tool offline with no opt-out and no actionable error), and
this report adds a fresh, fully-diagnosed macOS data point to that
pile:

- #1700 "dynamically fallback to getaddrinfo if hickory-dns doesn't
  work" — would have prevented this entirely.
- #1699 "Support --dns and env for selecting dns backend" — would
  give the user a workaround.
- #1708 "Download broken with VPN" — same surface, different
  trigger (`directive at line 0 is not recognized`).
- #1692 "Not working in Manjaro Linux on WiFi" — same surface,
  empty `/etc/resolv.conf`.
- #1698 "Cargo Binstall breaks with Safing Portmaster running" —
  same surface, DNS filtering trigger.

**To Reproduce**

Any macOS machine on a network where the router/Pi-hole/OpenWrt
advertises IPv6 DNS via Router Advertisement (RDNSS) reproduces this.
On such a machine, `scutil --dns` will show at least one nameserver
of the form `fe80::…%enN`, e.g.:

```
resolver #1
  search domain[0] : lan
  nameserver[0] : <REDACTED-RFC1918-IPV4>
  nameserver[1] : <REDACTED-IPV6-ULA>
  nameserver[2] : fe80::<REDACTED>%en1
  if_index : 15 (en1)
```

Then:

```sh
cargo binstall just-lsp --dry-run --no-confirm --log-level trace
```

…hangs for several seconds and exits with the error above. With
trace logging the actual failure chain becomes visible (see the
"Additional context" section).

A minimal Rust reproducer for the underlying hickory parse failure
is in
[hickory-dns#3667](https://github.com/hickory-dns/hickory-dns/issues/3667).
For cargo-binstall specifically, no separate repro is needed — any
release binary on a macOS host with the SC-store condition above
fails consistently.

**Expected behavior**

In order of preference:

1. **#1700**: when hickory fails to load system config, fall back to
   the OS resolver (`getaddrinfo` on Unix, `GetAddrInfoEx` on
   Windows) for normal A/AAAA lookups. The OS resolver on macOS
   handles link-local nameservers, VPN-injected resolvers, Portmaster,
   etc. correctly, by definition. This single change would resolve
   the macOS case in this issue, and based on the body of #1708,
   #1692, #1698 it would resolve those too.
2. **#1699**: ship a CLI flag and/or env var that lets the user
   force a specific DNS backend (`--dns=system` / `--dns=hickory`).
   This is the explicit-opt-out fallback and is independent of
   whether #1700 lands.
3. **Better error reporting** independent of either: when the hickory
   fallback chain is exercised, surface that fact in the user-visible
   error (e.g. "system DNS configuration could not be loaded; tried
   public DoH3 fallback (Quad9, Cloudflare, Google) which also
   failed"). The current "`dns error → request timed out`" message
   gives the user nothing to act on — they have to reach for
   `--log-level trace` and parse `hickory_*` spans to find out
   anything is wrong. Reporting `hickory failed to load system DNS
   config` at INFO/WARN level would let users diagnose this without
   filing an issue.

Most of the pain in this class of bug isn't the underlying hickory
behaviour — bugs happen — it's that cargo-binstall doesn't tell
users that hickory failed, and gives them no recourse when it does.

**System:**
- OS: macOS 26.3.1 (build 25D2128)
- Architecture: aarch64 (Apple Silicon)
- cargo-binstall: 1.19.1 (installed via Homebrew)
- rustc: 1.95.0 (59807616e 2026-04-14)

The affected machine reaches this state through a completely vanilla
setup: a Raspberry Pi running Pi-hole on the LAN, with IPv6 RA
RDNSS enabled (default in many Pi-hole / OpenWrt / dnsmasq setups).
No VPN, no firewall, no proxy, no manual DNS config — Pi-hole and
the Mac are talking via standard SLAAC/RA.

**Additional context**

End-user-visible failure from `cargo install-update -a` (which is
how the user originally hit this; it invokes `cargo-binstall`):

```
 INFO resolve: Resolving package: 'just-lsp@=0.4.4'
 INFO resolve:fetch_crate_matched{...}:do_send_request{...}: Received
   timeout error from reqwest. Delay future request by 200ms
[two more retries, same message]
ERROR Fatal error:
  × For crate just-lsp: could not GET https://index.crates.io/config.json:
  │ error sending request for url (https://index.crates.io/config.json)
  ├─▶ error sending request for url (https://index.crates.io/config.json)
  ├─▶ client error (Connect)
  ├─▶ dns error
  ╰─▶ request timed out
```

Trace-level excerpt from `cargo binstall --log-level trace just-lsp`
on the same machine (heavily abbreviated, full log available on
request):

```
DEBUG do_send_request{...}: starting new connection 'Some("index.crates.io")'
DEBUG do_send_request{...}:new_resolver: Using system DNS resolver configuration
DEBUG do_send_request{...}:new_resolver: hickory-dns: failed to load system DNS
   configuration; falling back to quad9, cloudflare and then googld:
   Msg("failed to parse nameserver address: invalid IP address syntax")
DEBUG do_send_request{...}:new_resolver: Resolver configuration complete
DEBUG do_send_request{...}: querying: index.crates.io. A
DEBUG do_send_request{...}: connecting config=NameServerConfig {
   ip: 2620:fe::fe, ..., connections: [ConnectionConfig { port: 443,
   protocol: H3 { server_name: "dns.quad9.net", ... } }] }
DEBUG do_send_request{...}: Ignoring error setting IP_RECVTOS on socket: ...
DEBUG do_send_request{...}:drive{id=0}: closing connection due to transport
   error: the cryptographic handshake failed: error 48: invalid peer
   certificate: UnknownIssuer
TRACE do_send_request{...}:drive{id=0}: connection closed
[same UnknownIssuer error against 2620:fe::9, then again against
 cloudflare-dns.com, then loss-detection timeouts on retries]
```

Independent verification that the network and `/etc/resolv.conf`
contents are fine:

```
$ dig +time=3 +tries=1 @<pihole-v4> index.crates.io A
;; Query time: 63 msec
;; ANSWER SECTION: ...

$ dig +time=3 +tries=1 @<pihole-v6-ula> index.crates.io A
;; Query time: 27 msec  (also works)

$ dig +time=3 +tries=1 "@fe80::<redacted>%en1" index.crates.io A
;; Query time: 23 msec  (link-local also works fine)

$ curl --max-time 10 https://index.crates.io/config.json
HTTP/2 200  (succeeds in ~50 ms)
```

So every layer except hickory's macOS system-config loader works.
Once that loader bails out on the `%en1` line and falls through to
DoH3, the second failure (UnknownIssuer) finishes the user off.

**Workaround**

For affected users on macOS, manually overriding DNS on the active
network service drops the link-local entry from `/etc/resolv.conf`,
which sidesteps the hickory parse failure:

```
networksetup -setdnsservers "Wi-Fi" <pihole-ipv4> 1.1.1.1
```

(Reverts with `-setdnsservers "Wi-Fi" empty`.)

This is a workaround, not a fix — many users won't think to look at
their DNS configuration when `cargo install-update` times out.

Hmmm im hitting some errors:

```sh
$ cargo install cargo-binstall --git https://github.com/cargo-bins/cargo-binstall --branch NobodyXu-patch-1 --locked

....

warning: ring@0.17.14: ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/a0330e891e733f4e-p256-nistz.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/p256-nistz.c"
error: failed to run custom build command for `ring v0.17.14`

Caused by:
  process didn't exit successfully: `/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-b09315f65db3d327/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=CARGO_MANIFEST_DIR
  cargo:rerun-if-env-changed=CARGO_PKG_NAME
  cargo:rerun-if-env-changed=CARGO_PKG_VERSION_MAJOR
  cargo:rerun-if-env-changed=CARGO_PKG_VERSION_MINOR
  cargo:rerun-if-env-changed=CARGO_PKG_VERSION_PATCH
  cargo:rerun-if-env-changed=CARGO_PKG_VERSION_PRE
  cargo:rerun-if-env-changed=CARGO_MANIFEST_LINKS
  cargo:rerun-if-env-changed=RING_PREGENERATE_ASM
  cargo:rerun-if-env-changed=OUT_DIR
  cargo:rerun-if-env-changed=CARGO_CFG_TARGET_ARCH
  cargo:rerun-if-env-changed=CARGO_CFG_TARGET_OS
  cargo:rerun-if-env-changed=CARGO_CFG_TARGET_ENV
  cargo:rerun-if-env-changed=CARGO_CFG_TARGET_ENDIAN
  cargo:rerun-if-env-changed=CC_aarch64-apple-darwin
  CC_aarch64-apple-darwin = None
  cargo:rerun-if-env-changed=CC_aarch64_apple_darwin
  CC_aarch64_apple_darwin = None
  cargo:rerun-if-env-changed=HOST_CC
  HOST_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  cargo:rerun-if-env-changed=MACOSX_DEPLOYMENT_TARGET
  MACOSX_DEPLOYMENT_TARGET = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=HOST_CFLAGS
  HOST_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64_apple_darwin
  CFLAGS_aarch64_apple_darwin = None
  cargo:rerun-if-env-changed=CFLAGS_aarch64-apple-darwin
  CFLAGS_aarch64-apple-darwin = None
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/curve25519/curve25519.c:22:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/mem.h:18:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.
  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/25ac62e5b3c53843-curve25519.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/curve25519/curve25519.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/aes/aes_nohw.c:15:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/aes.h:18:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/0bbbd18bda93c05b-aes_nohw.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/aes/aes_nohw.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/bn/montgomery.c:15:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/bn/internal.h:19:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/00c879ee3285a50d-montgomery.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/bn/montgomery.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/bn/montgomery_inv.c:15:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/bn/internal.h:19:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/00c879ee3285a50d-montgomery_inv.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/bn/montgomery_inv.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/ecp_nistz.c:15:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/ecp_nistz.h:18:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/a0330e891e733f4e-ecp_nistz.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/ecp_nistz.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/gfp_p256.c:15:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/./p256_shared.h:27:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/a0330e891e733f4e-gfp_p256.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/gfp_p256.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/gfp_p384.c:15:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/../../limbs/limbs.h:18:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/a0330e891e733f4e-gfp_p384.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/gfp_p384.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/p256.c:19:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/a0330e891e733f4e-p256.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/p256.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/limbs/limbs.c:15:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/limbs/limbs.h:18:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/aaa1ba3e455ee2e1-limbs.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/limbs/limbs.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/mem.c:15:
  cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/mem.h:18:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/a4019cc0736b0423-mem.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/mem.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/poly1305/poly1305.c:19:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/d5a9841f3dc6e253-poly1305.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/poly1305/poly1305.c"cargo:warning=In file included from /Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/p256-nistz.c:24:
  cargo:warning=/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include/ring-core/base.h:33:10: fatal error: 'TargetConditionals.h' file not found
  cargo:warning=   33 | #include <TargetConditionals.h>
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=1 error generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/a0330e891e733f4e-p256-nistz.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/p256-nistz.c"
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0
  exit status: 0

  --- stderr


  error occurred in cc-rs: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL="C" "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=arm64-apple-macosx" "-mmacosx-version-min=26.5" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/include" "-I" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/pregenerated" "-Wall" "-Wextra" "-fvisibility=hidden" "-std=c1x" "-Wall" "-Wbad-function-cast" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wnested-externs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wstrict-prototypes" "-Wundef" "-Wuninitialized" "-gfull" "-DNDEBUG" "-o" "/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem/release/build/ring-592787916c0bc61f/out/a0330e891e733f4e-p256-nistz.o" "-c" "/Users/compux72/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ring-0.17.14/crypto/fipsmodule/ec/p256-nistz.c"


error: failed to compile `cargo-binstall v1.19.1 (https://github.com/cargo-bins/cargo-binstall?branch=NobodyXu-patch-1#845d84ed)`, intermediate artifacts can be found at `/var/folders/jf/82bk2rgx1p75514p70njztnh0000gn/T/cargo-installqRxxem`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_BUILD_BUILD_DIR` to that path.
```


How do I disable ring?

--- a/crates/binstalk-downloader/src/remote/resolver.rs
+++ b/crates/binstalk-downloader/src/remote/resolver.rs
@@ -45,14 +45,14 @@ fn get_system_configs() -> (ResolverConfig, ResolverOpts) {
     system_conf::read_system_conf().unwrap_or_else(|err| {
         debug!(
             "hickory-dns: failed to load system DNS configuration; \
-            falling back to quad9, cloudflare and then googld: {:?}",
+            falling back to cloudflare and then googld: {:?}",
             err
         );
 
         let mut config = ResolverConfig::default();
         let mut opts = ResolverOpts::default();
 
-        let dns_providers = [QUAD9, CLOUDFLARE, GOOGLE];
+        let dns_providers = [CLOUDFLARE, GOOGLE];
         // quic first as it is secure while being the fastes
         dns_providers
             .iter()

